pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,477 @@
+from pennylane import math as qmlmath
+from pennylane import operation
+from pennylane.queuing import QueuingContext
+from pennylane.wires import Wires
+
+
+# pylint: disable=no-member
+class ControlledOperation(operation.Operation):
+    """Operation-specific methods and properties for the ``Controlled`` class.
+
+    Dynamically mixed in based on the provided base operator.  If the base operator is an
+    Operation, this class will be mixed in.
+
+    When we no longer rely on certain functionality through `Operation`, we can get rid of this
+    class.
+
+    Defers inversion behavior to base.  This way we don't have to modify the ``Controlled.matrix``
+    and ``Controlled.eigvals`` to account for in-place inversion. In-place inversion of a matrix
+    """
+
+    @property
+    def _inverse(self):
+        return False
+
+    @_inverse.setter
+    def _inverse(self, boolean):
+        self.base._inverse = boolean  # pylint: disable=protected-access
+        # refresh name as base_name got updated.
+        self._name = f"C{self.base.name}"
+
+    def inv(self):
+        self.base.inv()
+        # refresh name as base_name got updated.
+        self._name = f"C{self.base.name}"
+        return self
+
+    @property
+    def base_name(self):
+        return f"C{self.base.base_name}"
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def grad_method(self):
+        return self.base.grad_method
+
+    # pylint: disable=missing-function-docstring
+    @property
+    def basis(self):
+        return self.base.basis
+
+    @property
+    def parameter_frequencies(self):
+        if self.base.num_params == 1:
+            try:
+                base_gen = qml.generator(self.base, format="observable")
+            except operation.GeneratorUndefinedError as e:
+                raise operation.ParameterFrequenciesUndefinedError(
+                    f"Operation {self.base.name} does not have parameter frequencies defined."
+                ) from e
+
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    action="ignore", message=r".+ eigenvalues will be computed numerically\."
+                )
+                base_gen_eigvals = qml.eigvals(base_gen)
+
+            # The projectors in the full generator add a eigenvsalue of `0` to
+            # the eigenvalues of the base generator.
+            gen_eigvals = np.append(base_gen_eigvals, 0)
+
+            processed_gen_eigvals = tuple(np.round(gen_eigvals, 8))
+            return [qml.gradients.eigvals_to_frequencies(processed_gen_eigvals)]
+        raise operation.ParameterFrequenciesUndefinedError(
+            f"Operation {self.name} does not have parameter frequencies defined, "
+            "and parameter frequencies can not be computed via generator for more than one parameter."
+        )
+
+
+# pylint: disable=too-many-arguments, too-many-public-methods
+class Controlled(operation.Operator):
```

Comment:
I was importing so many things from the `operation` module I just decided to import the module as a whole instead of everything I was using from it individually. Any reason I should individually import `Operator`?